### PR TITLE
Implement GraphQL endpoint for GetMetricHistoryBulkInterval

### PR DESCRIPTION
**Pull request by daniellok-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/daniellok-db/mlflow/pull/12584?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12584/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12584
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

This PR follows [this guide](https://github.com/mlflow/mlflow/blob/14577103c15b243441e951f0e84891f13efaea4f/dev/proto_to_graphql/README.md#how-to-onboard-a-new-rpc-to-graphql) on migrating endpoints to GraphQL. This is the first of probably many PRs for this migration. To start, I'm migrating `GetMetricHistoryBulkInterval`, which is the endpoint that powers data sampling in charts.

I'm doing two things:
1. Adding a graphQL endpoint that matches the RPC (`mlflowGetMetricHistoryBulkInterval`)
2. Adding a convenience resolver on `mlflowGetRun` that allows us to query metrics for a run. Perhaps it might be also be useful to add this on the experiment level.

Both these use-cases come with a test

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Added two new unit tests to test these endpoints

Also tested manually in GraphQL playground:

https://github.com/mlflow/mlflow/assets/148037680/51d783f6-bb6c-42cf-a3b9-1770ce1ddb44

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/server/graphql/graphql_schema_extensions.py [hunk 1]
```diff
@@ -54,6 +59,16 @@ def resolve_model_versions(self, info):
         parse_dict(input_dict, request_message)
         return mlflow.server.handlers.search_model_versions_impl(request_message).model_versions
 
+    def resolve_metric_history(self, info, metricKey):
```
Comment (hubertzub-db) [inline mlflow/server/graphql/graphql_schema_extensions.py:62]: Disclaimer: just thinking out loud 😅 
While it's certainly convenient, I wonder if we'll ever have a use case for fetching sampled metrics history alongside run data (in the same query). After all, all existing scenarios for fetching this data involve lazily loading it for charts, which happens independently. I'm totally fine with introducing nice extras, I just wonder if this change won't add something never to be used but still requiring maintenance (tests, docs etc)

**Review by hubertzub-db — APPROVED** [review]
left one ugly whine 😅 otherwise it looks great! I think it'd be good to have @edwardfeng-db review it as well

File: mlflow/server/graphql/graphql_schema_extensions.py (hunk 1 above)
Comment (daniellok-db) [reply mlflow/server/graphql/graphql_schema_extensions.py:62]: Makes sense! I'll remove this for now, I was mainly implementing this to make sure I understood the whole query joining thing haha. I agree that there isn't going to be a use-case for this.

File: tests/tracking/test_rest_tracking.py [hunk 2]
```diff
@@ -2220,3 +2220,41 @@ def test_get_trace_artifact_handler(mlflow_client):
     # Validate content
     trace_data = TraceData.from_dict(json.loads(response.text))
     assert trace_data.spans[0].to_dict() == span.to_dict()
+
+
+def test_get_metric_history_bulk_interval_grapql(mlflow_client):
```
Comment (BenWilson2) [inline tests/tracking/test_rest_tracking.py:2225]:
```suggestion
def test_get_metric_history_bulk_interval_graphql(mlflow_client):
```

Comment (daniellok-db) [reply tests/tracking/test_rest_tracking.py:2225]: thanks for the catch!

**Review by edwardfeng-db — APPROVED** [review]

**Pull request merged**